### PR TITLE
Add Observer

### PR DESCRIPTION
**Pull request by CyberDrudge**
Add a basic observer and its context

### CODE REVIEW

File: src/main/java/io/appform/dropwizard/actors/observers/ObserverContext.java [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+
+import lombok.Builder;
+import lombok.Value;
+
+@Value
+@Builder
+public class ObserverContext {
+    String opType;
+    String queueName;
+}
```
Comment (CyberDrudge) [inline src/main/java/io/appform/dropwizard/actors/observers/ObserverContext.java:11]: Should we add some details about queue shard (shardId or something else) or sideline queue?

File: src/main/java/io/appform/dropwizard/actors/observers/RMQPublishObserver.java [hunk 2]
```diff
@@ -0,0 +1,28 @@
+package io.appform.dropwizard.actors.observers;
+
+import lombok.Getter;
+
+import java.util.function.Supplier;
+
+/**
+ *
+ */
+public abstract class RMQPublishObserver {
```
Comment (Kanikakhetawat) [inline src/main/java/io/appform/dropwizard/actors/observers/RMQPublishObserver.java:10]: how are the observers getting called?

File: src/main/java/io/appform/dropwizard/actors/observers/ObserverContext.java [hunk 3]
```diff
@@ -0,0 +1,11 @@
+package io.appform.dropwizard.actors.observers;
+
+import lombok.Builder;
+import lombok.Value;
+
+@Value
+@Builder
+public class ObserverContext {
```
Comment (Kanikakhetawat) [inline src/main/java/io/appform/dropwizard/actors/observers/ObserverContext.java:8]: have different contexts for publish and consume

File: src/main/java/io/appform/dropwizard/actors/config/RMQConfig.java [hunk 4]
```diff
@@ -89,4 +89,6 @@ public class RMQConfig {
     @Valid
     private List<ConnectionConfig> connections;
 
+    @Builder.Default
+    private boolean metricsEnabled = false;
```
Comment (Kanikakhetawat) [inline src/main/java/io/appform/dropwizard/actors/config/RMQConfig.java:93]: publishMetricsEnabled

File: src/main/java/io/appform/dropwizard/actors/connectivity/RMQConnection.java [hunk 5]
```diff
@@ -249,4 +254,8 @@ private Map<String, Object> getPriorityOpts(final ActorConfig actorConfig) {
         }
     }
 
+    private void setupObservers(final RMQConfig config,
```
Comment (Kanikakhetawat) [inline src/main/java/io/appform/dropwizard/actors/connectivity/RMQConnection.java:257]: The actual publish method should also be called through an observer. There can be a chain of observers, that sequence needs to be created in this method

File: src/main/java/io/appform/dropwizard/actors/connectivity/RMQConnection.java [hunk 6]
```diff
@@ -129,6 +133,7 @@ public void handleUnblocked() {
             }
         });
         channel = connection.createChannel();
+        setupObservers(config, environment.metrics());
```
Comment (Kanikakhetawat) [inline src/main/java/io/appform/dropwizard/actors/connectivity/RMQConnection.java:136]: Add a method to register publish observer from outside

File: src/main/java/io/appform/dropwizard/actors/metrics/MetricKeyData.java [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+import lombok.Builder;
+import lombok.Data;
+import lombok.EqualsAndHashCode;
+import lombok.NoArgsConstructor;
+
+@NoArgsConstructor
+@AllArgsConstructor
+@Builder
+@Data
+@EqualsAndHashCode
```
Comment (Kanikakhetawat) [inline src/main/java/io/appform/dropwizard/actors/metrics/MetricKeyData.java:13]: Remove this. Also use @Value

File: src/main/java/io/appform/dropwizard/actors/metrics/MetricObserver.java [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+import lombok.val;
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Supplier;
+
+
+@Slf4j
+public class MetricObserver extends RMQPublishObserver {
```
Comment (Kanikakhetawat) [inline src/main/java/io/appform/dropwizard/actors/metrics/MetricObserver.java:20]: PublishMetricObserver

File: src/main/java/io/appform/dropwizard/actors/metrics/MetricObserver.java [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+
+    public MetricObserver(final RMQConfig rmqConfig,
+                          final MetricRegistry metricRegistry) {
+        this.rmqConfig = rmqConfig;
+        this.metricRegistry = metricRegistry;
+    }
+
+    @Override
+    public <T> T execute(ObserverContext context, Supplier<T> supplier) {
+        if (!rmqConfig.isMetricsEnabled()) {
```
Comment (Kanikakhetawat) [inline src/main/java/io/appform/dropwizard/actors/metrics/MetricObserver.java:35]: give a capability to enable metrics per queue

File: src/main/java/io/appform/dropwizard/actors/metrics/MetricUtil.java [hunk 10]
```diff
@@ -0,0 +1,28 @@
+package io.appform.dropwizard.actors.metrics;
+
+import lombok.experimental.UtilityClass;
+import lombok.val;
+
+@UtilityClass
+public class MetricUtil {
+
+    private static final String RMQ_PREFIX = "rmqconnection";
```
Comment (Kanikakhetawat) [inline src/main/java/io/appform/dropwizard/actors/metrics/MetricUtil.java:9]: change this and remove connection from prefix

**Review by Kanikakhetawat — CHANGES_REQUESTED** [review]

File: src/main/java/io/appform/dropwizard/actors/base/UnmanagedPublisher.java [hunk 11]
```diff
@@ -42,6 +44,7 @@ public UnmanagedPublisher(
         this.connection = connection;
         this.mapper = mapper;
         this.queueName = NamingUtils.queueName(config.getPrefix(), name);
+        this.observer = connection.getRootObserver();
```
Comment (Kanikakhetawat) [inline src/main/java/io/appform/dropwizard/actors/base/UnmanagedPublisher.java:47]: How are we chaining the observers if we are just calling the root observer here?

File: src/main/java/io/appform/dropwizard/actors/metrics/MetricUtil.java [hunk 12]
```diff
@@ -0,0 +1,40 @@
+package io.appform.dropwizard.actors.metrics;
+
+import io.appform.dropwizard.actors.config.MetricConfig;
+import lombok.experimental.UtilityClass;
+import lombok.val;
+
+@UtilityClass
+public class MetricUtil {
+
+    private static final String RMQ_PREFIX = "rmq";
```
Comment (Kanikakhetawat) [inline src/main/java/io/appform/dropwizard/actors/metrics/MetricUtil.java:10]: will same be used for both publish and consume?

File: src/main/java/io/appform/dropwizard/actors/base/UnmanagedPublisher.java [hunk 13]
```diff
@@ -76,26 +94,38 @@ public final void publishWithExpiry(final Message message, final long expiryInMs
                     .headers(ImmutableMap.of(MESSAGE_EXPIRY_TEXT, expiresAt))
                     .build();
         }
-        publish(message, properties);
+        publish(message, properties, RMQOperations.PUBLISH_WITH_EXPIRY.name());
     }
 
     public final void publish(final Message message) throws Exception {
-        publish(message, MessageProperties.MINIMAL_PERSISTENT_BASIC);
+        publish(message, MessageProperties.MINIMAL_PERSISTENT_BASIC, RMQOperations.PUBLISH.name());
```
Comment (Kanikakhetawat) [inline src/main/java/io/appform/dropwizard/actors/base/UnmanagedPublisher.java:101]: pass the enum itself. No need to do .name()

File: src/main/java/io/appform/dropwizard/actors/actor/UnmanagedBaseActor.java [hunk 14]
```diff
@@ -127,7 +128,7 @@ public final void publish(final Message message) throws Exception {
     }
 
     public final void publish(final Message message, final AMQP.BasicProperties properties) throws Exception {
-        publishActor().publish(message, properties);
+        publishActor().publish(message, properties, RMQOperations.PUBLISH.name());
```
Comment (Kanikakhetawat) [inline src/main/java/io/appform/dropwizard/actors/actor/UnmanagedBaseActor.java:131]: pass the enum

File: src/main/java/io/appform/dropwizard/actors/observers/TerminalObserver.java [hunk 15]
```diff
@@ -0,0 +1,22 @@
+package io.appform.dropwizard.actors.observers;
+
+import java.util.function.Supplier;
+
+/**
+ *
+ */
+public final class TerminalObserver extends RMQObserver {
```
Comment (Kanikakhetawat) [inline src/main/java/io/appform/dropwizard/actors/observers/TerminalObserver.java:8]: TerminalRMQObserver

File: src/main/java/io/appform/dropwizard/actors/metrics/RMQMetricObserver.java [hunk 16]
```diff
@@ -0,0 +1,94 @@
...
+import lombok.val;
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Supplier;
+
+
+@Slf4j
+public class RMQMetricObserver extends RMQObserver {
```
Comment (Kanikakhetawat) [inline src/main/java/io/appform/dropwizard/actors/metrics/RMQMetricObserver.java:20]: add java doc and sample metrics in comments

File: src/main/java/io/appform/dropwizard/actors/metrics/MetricUtil.java (hunk 12 above)
Comment (Kanikakhetawat) [reply src/main/java/io/appform/dropwizard/actors/metrics/MetricUtil.java:10]: maybe use rmq.actor

File: src/main/java/io/appform/dropwizard/actors/metrics/MetricKeyData.java [hunk 17]
```diff
@@ -0,0 +1,13 @@
+package io.appform.dropwizard.actors.metrics;
+
+import lombok.Builder;
+import lombok.Data;
+import lombok.Value;
+
+@Builder
+@Data
```
Comment (Kanikakhetawat) [inline src/main/java/io/appform/dropwizard/actors/metrics/MetricKeyData.java:8]: remove this

File: src/main/java/io/appform/dropwizard/actors/connectivity/RMQConnection.java [hunk 18]
```diff
@@ -59,17 +62,22 @@ public class RMQConnection implements Managed {
...
+    @Getter
+    private final RMQObserver rootObserver;
+
 
     public RMQConnection(final String name,
                          final RMQConfig config,
                          final ExecutorService executorService,
                          final Environment environment,
-                         final TtlConfig ttlConfig) {
+                         final TtlConfig ttlConfig, RMQObserver rootObserver) {
```
Comment (Kanikakhetawat) [inline src/main/java/io/appform/dropwizard/actors/connectivity/RMQConnection.java:74]: move the arg to next line

File: src/main/java/io/appform/dropwizard/actors/base/UnmanagedPublisher.java [hunk 19]
```diff
@@ -51,18 +56,31 @@ public final void publishWithDelay(final Message message, final long delayMillis
...
         if (config.getDelayType() == DelayType.TTL) {
-            publishChannel.basicPublish(ttlExchange(config),
-                    queueName,
-                    new AMQP.BasicProperties.Builder()
-                            .expiration(String.valueOf(delayMilliseconds))
-                            .deliveryMode(2)
-                            .build(),
-                    mapper().writeValueAsBytes(message));
+            val context = ObserverContext.builder()
+                    .operation(RMQOperations.PUBLISH_WITH_DELAY.name())
```
Comment (Kanikakhetawat) [inline src/main/java/io/appform/dropwizard/actors/base/UnmanagedPublisher.java:60]: is the operation correct here? this has just specified a ttl but there is no delay

File: src/main/java/io/appform/dropwizard/actors/base/UnmanagedPublisher.java [hunk 20]
```diff
@@ -51,18 +56,31 @@ public final void publishWithDelay(final Message message, final long delayMillis
...
+                    publishChannel.basicPublish(ttlExchange(config),
+                            queueName,
+                            new AMQP.BasicProperties.Builder()
+                                    .expiration(String.valueOf(delayMilliseconds))
+                                    .deliveryMode(2)
+                                    .headers(headers)
+                                    .build(),
+                            mapper().writeValueAsBytes(message));
+                } catch (IOException e) {
+                    e.printStackTrace();
```
Comment (Kanikakhetawat) [inline src/main/java/io/appform/dropwizard/actors/base/UnmanagedPublisher.java:75]: throw a proper exception and remove printStackTrace

File: src/main/java/io/appform/dropwizard/actors/base/UnmanagedPublisher.java [hunk 21]
```diff
@@ -76,26 +94,38 @@ public final void publishWithExpiry(final Message message, final long expiryInMs
...
-        val enrichedProperties = getEnrichedProperties(properties);
-        publishChannel.basicPublish(config.getExchange(), routingKey, enrichedProperties, mapper().writeValueAsBytes(message));
+        val context = ObserverContext.builder().operation(operation).queueName(routingKey).build();
+        observer.executePublish(context, () -> {
+            val headers = context.getHeaders();
+            val enrichedProperties = getEnrichedProperties(properties, headers);
+            try {
+                publishChannel.basicPublish(config.getExchange(), routingKey, enrichedProperties, mapper().writeValueAsBytes(message));
+            } catch (IOException e) {
+                e.printStackTrace();
```
Comment (Kanikakhetawat) [inline src/main/java/io/appform/dropwizard/actors/base/UnmanagedPublisher.java:118]: throw a proper exception and remove printStackTrace

File: src/main/java/io/appform/dropwizard/actors/base/UnmanagedPublisher.java [hunk 22]
```diff
@@ -51,18 +56,31 @@ public final void publishWithDelay(final Message message, final long delayMillis
         }
 
         if (config.getDelayType() == DelayType.TTL) {
-            publishChannel.basicPublish(ttlExchange(config),
-                    queueName,
```
Comment (Kanikakhetawat) [inline src/main/java/io/appform/dropwizard/actors/base/UnmanagedPublisher.java:55]: this should be shardedQueueName if sharding is done. Check if this is an issue with existing code

File: src/main/java/io/appform/dropwizard/actors/base/UnmanagedPublisher.java [hunk 23]
```diff
@@ -76,26 +94,38 @@ public final void publishWithExpiry(final Message message, final long expiryInMs
...
+    public final void publish(final Message message, final AMQP.BasicProperties properties, final String operation) throws Exception {
         String routingKey;
         if (config.isSharded()) {
             routingKey = NamingUtils.getShardedQueueName(queueName, getShardId());
         } else {
             routingKey = queueName;
         }
-        val enrichedProperties = getEnrichedProperties(properties);
-        publishChannel.basicPublish(config.getExchange(), routingKey, enrichedProperties, mapper().writeValueAsBytes(message));
+        val context = ObserverContext.builder().operation(operation).queueName(routingKey).build();
```
Comment (Kanikakhetawat) [inline src/main/java/io/appform/dropwizard/actors/base/UnmanagedPublisher.java:111]: sending shardedQueueName might be an issue, all the metrics will get changed if sharding is enabled. Revisit this

File: src/main/java/io/appform/dropwizard/actors/base/UnmanagedConsumer.java [hunk 24]
```diff
@@ -59,21 +63,25 @@ public UnmanagedConsumer(final String name,
...
             if (config.isSharded()) {
                 queueNameForConsumption = NamingUtils.getShardedQueueName(queueName, i % config.getShardCount());
             } else {
                 queueNameForConsumption = queueName;
             }
 
+            final Handler<Message> handler =
+                    new Handler<>(consumeChannel, mapper, clazz, prefetchCount, errorCheckFunction, retryStrategy,
+                                  exceptionHandler, handlerFunction, expiredMessageHandlingFunction, observer,
+                            queueNameForConsumption);
```
Comment (Kanikakhetawat) [inline src/main/java/io/appform/dropwizard/actors/base/UnmanagedConsumer.java:83]: revisit the sharded queue name

File: src/main/java/io/appform/dropwizard/actors/base/Handler.java [hunk 25]
```diff
@@ -61,15 +71,25 @@ public Handler(final Channel channel,
         this.expiredMessageHandlingFunction = expiredMessageHandlingFunction;
     }
 
-    private boolean handle(final Message message, final MessageMetadata messageMetadata, final boolean expired) throws Exception {
+    private boolean handle(final Message message, final MessageMetadata messageMetadata, final boolean expired, final Map<String, Object> headers) throws Exception {
```
Comment (Kanikakhetawat) [inline src/main/java/io/appform/dropwizard/actors/base/Handler.java:74]: move arguments to separate lines

File: src/main/java/io/appform/dropwizard/actors/base/Handler.java [hunk 26]
```diff
@@ -61,15 +71,25 @@ public Handler(final Channel channel,
...
+                .queueName(queueName)
+                .headers(headers)
+                .build();
+        return observer.executeConsume(context, () -> {
+            try {
+                return expired
+                        ? expiredMessageHandlingFunction.apply(message, messageMetadata)
+                        : messageHandlingFunction.apply(message, messageMetadata);
+            } catch (Exception e) {
+                e.printStackTrace();
```
Comment (Kanikakhetawat) [inline src/main/java/io/appform/dropwizard/actors/base/Handler.java:87]: fix this, throw proper exception

File: src/main/java/io/appform/dropwizard/actors/observers/ObserverContext.java [hunk 27]
```diff
@@ -0,0 +1,17 @@
...
+
+import io.appform.dropwizard.actors.common.RMQOperation;
+import lombok.Builder;
+import lombok.Data;
+import lombok.Setter;
+
+import java.util.Map;
+
+@Data
+@Builder
```
Comment (Kanikakhetawat) [inline src/main/java/io/appform/dropwizard/actors/observers/ObserverContext.java:11]: use @Value and remove @Data and @Setter

File: src/main/java/io/appform/dropwizard/actors/base/UnmanagedConsumer.java [hunk 28]
```diff
@@ -59,21 +63,24 @@ public UnmanagedConsumer(final String name,
...
-                                  exceptionHandler, handlerFunction, expiredMessageHandlingFunction);
+
             String queueNameForConsumption;
             if (config.isSharded()) {
                 queueNameForConsumption = NamingUtils.getShardedQueueName(queueName, i % config.getShardCount());
             } else {
                 queueNameForConsumption = queueName;
             }
 
+            final Handler<Message> handler =
```
Comment (Kanikakhetawat) [inline src/main/java/io/appform/dropwizard/actors/base/UnmanagedConsumer.java:80]: revert this ordering change

File: src/test/java/io/appform/dropwizard/actors/metrics/RMQMetricObserverTest.java [hunk 29]
```diff
@@ -0,0 +1,87 @@
...
+import io.appform.dropwizard.actors.config.RMQConfig;
+import io.appform.dropwizard.actors.observers.ObserverContext;
+import lombok.val;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
```
Comment (Kanikakhetawat) [inline src/test/java/io/appform/dropwizard/actors/metrics/RMQMetricObserverTest.java:15]: write some tests to validate the chaining of observers

**Review by Kanikakhetawat — CHANGES_REQUESTED** [review]

File: src/main/java/io/appform/dropwizard/actors/base/UnmanagedPublisher.java (hunk 22 above)
Comment (CyberDrudge) [reply src/main/java/io/appform/dropwizard/actors/base/UnmanagedPublisher.java:55]: Added sharded queue name

File: src/main/java/io/appform/dropwizard/actors/base/UnmanagedPublisher.java [hunk 30]
```diff
@@ -51,18 +60,31 @@ public final void publishWithDelay(final Message message, final long delayMillis
...
 
         if (config.getDelayType() == DelayType.TTL) {
-            publishChannel.basicPublish(ttlExchange(config),
-                    queueName,
-                    new AMQP.BasicProperties.Builder()
-                            .expiration(String.valueOf(delayMilliseconds))
-                            .deliveryMode(2)
-                            .build(),
-                    mapper().writeValueAsBytes(message));
+            String routingKey = getRoutingKey();
```
Comment (Kanikakhetawat) [inline src/main/java/io/appform/dropwizard/actors/base/UnmanagedPublisher.java:63]: use val

File: src/main/java/io/appform/dropwizard/actors/base/UnmanagedPublisher.java [hunk 31]
```diff
@@ -76,22 +98,26 @@ public final void publishWithExpiry(final Message message, final long expiryInMs
...
-        String routingKey;
-        if (config.isSharded()) {
-            routingKey = NamingUtils.getShardedQueueName(queueName, getShardId());
-        } else {
-            routingKey = queueName;
-        }
-        val enrichedProperties = getEnrichedProperties(properties);
-        publishChannel.basicPublish(config.getExchange(), routingKey, enrichedProperties, mapper().writeValueAsBytes(message));
+    public final void publish(final Message message, final AMQP.BasicProperties properties, final RMQOperation operation) throws Exception {
+        String routingKey = getRoutingKey();
```
Comment (Kanikakhetawat) [inline src/main/java/io/appform/dropwizard/actors/base/UnmanagedPublisher.java:109]: routing key should only be fetched if sharding is enabled for the queue

**Review by Kanikakhetawat — CHANGES_REQUESTED** [review]

File: src/main/java/io/appform/dropwizard/actors/base/UnmanagedPublisher.java (hunk 31 above)
Comment (CyberDrudge) [reply src/main/java/io/appform/dropwizard/actors/base/UnmanagedPublisher.java:109]: getRoutingKey method checks `config.isSharded()` and returns sharded queue name or queue name accordingly.

File: src/main/java/io/appform/dropwizard/actors/RabbitmqActorBundle.java [hunk 32]
```diff
@@ -78,4 +88,25 @@ public RMQConnection getConnection() {
...
+        this.observers.add(observer);
+        log.info("Registered observer: " + observer.getClass().getSimpleName());
+    }
+
+    private void setupObservers(final MetricRegistry metricRegistry) {
+        //Terminal observer calls the actual method
+        rootObserver = new TerminalRMQObserver();
+        for (var observer : observers) {
+            if (null == observer) {
+                return;
```
Comment (santanusinha) [inline src/main/java/io/appform/dropwizard/actors/RabbitmqActorBundle.java:105]: this should continue to next. Not return?

File: src/main/java/io/appform/dropwizard/actors/base/UnmanagedConsumer.java [hunk 33]
```diff
@@ -59,14 +63,15 @@ public UnmanagedConsumer(final String name,
         this.queueName = NamingUtils.queueName(config.getPrefix(), name);
         this.retryStrategy = retryStrategyFactory.create(config.getRetryConfig());
         this.exceptionHandler = exceptionHandlingFactory.create(config.getExceptionHandlerConfig());
+        this.observer = connection.getRootObserver();
```
Comment (santanusinha) [inline src/main/java/io/appform/dropwizard/actors/base/UnmanagedConsumer.java:66]: Ensure not null, otherwise this will break during runtime

**Review by santanusinha — CHANGES_REQUESTED** [review]

File: src/main/java/io/appform/dropwizard/actors/base/UnmanagedConsumer.java (hunk 33 above)
Comment (CyberDrudge) [reply src/main/java/io/appform/dropwizard/actors/base/UnmanagedConsumer.java:66]: Added Observer null check in RabbitmqActorBundle.java

File: src/main/java/io/appform/dropwizard/actors/common/RMQOperation.java [hunk 34]
```diff
@@ -0,0 +1,8 @@
+package io.appform.dropwizard.actors.common;
+
+public enum RMQOperation {
+    CONSUME,
+    PUBLISH,
```
Comment (r0goyal) [inline src/main/java/io/appform/dropwizard/actors/common/RMQOperation.java:5]: Single publish operation should be required. No need for publish with different kind of attributes

File: src/main/java/io/appform/dropwizard/actors/metrics/MetricData.java [hunk 35]
```diff
@@ -0,0 +1,15 @@
...
+
+import com.codahale.metrics.Meter;
+import com.codahale.metrics.Timer;
+import lombok.Builder;
+import lombok.Value;
+
+@Value
+@Builder
+public class MetricData {
+    Meter total;
```
Comment (r0goyal) [inline src/main/java/io/appform/dropwizard/actors/metrics/MetricData.java:11]: add metric for message redelivery

File: src/main/java/io/appform/dropwizard/actors/metrics/RMQMetricObserver.java [hunk 36]
```diff
@@ -0,0 +1,129 @@
...
+    @Override
+    public <T> T executeConsume(final ConsumeObserverContext context, final Supplier<T> supplier) {
+        if (!MetricUtil.isMetricApplicable(rmqConfig.getMetricConfig(), context.getQueueName())) {
+            return proceedConsume(context, supplier);
+        }
+        val isRedelivered = context.isRedelivered();
+        val metricData = getMetricData(context);
+        val metricDataForRedelivery = isRedelivered ? getMetricDataForRedelivery(context) : null;
+        metricData.getTotal().mark();
+        if (isRedelivered) {
```
Comment (Kanikakhetawat) [inline src/main/java/io/appform/dropwizard/actors/metrics/RMQMetricObserver.java:66]: the null check should be done on metricDataForRedelivery

File: src/main/java/io/appform/dropwizard/actors/metrics/RMQMetricObserver.java [hunk 37]
```diff
@@ -0,0 +1,129 @@
...
+            metricDataForRedelivery.getTotal().mark();
+        }
+        val timer = metricData.getTimer().time();
+        val redeliveryTimer =  isRedelivered ? metricDataForRedelivery.getTimer().time(): null;
+        try {
+            val response = proceedConsume(context, supplier);
+            metricData.getSuccess().mark();
+            if (isRedelivered) {
```
Comment (Kanikakhetawat) [inline src/main/java/io/appform/dropwizard/actors/metrics/RMQMetricObserver.java:74]: check redeliveryTimer here

File: src/test/java/io/appform/dropwizard/actors/metrics/RMQMetricObserverTest.java [hunk 38]
```diff
@@ -0,0 +1,84 @@
...
+                .operation(RMQOperation.PUBLISH)
+                .queueName("default")
+                .build();
+        assertEquals(terminate(), rmqMetricObserver.executePublish(context, this::terminate));
+        val key = MetricKeyData.builder().operation(context.getOperation()).queueName(context.getQueueName()).build();
+        validateMetrics(rmqMetricObserver.getMetricCache().get(key), 1, 0);
+    }
+
+    @Test
+    void testExecuteWithException() {
```
Comment (Kanikakhetawat) [inline src/test/java/io/appform/dropwizard/actors/metrics/RMQMetricObserverTest.java:58]: add test cases for redelivered metrics

File: src/main/java/io/appform/dropwizard/actors/metrics/MetricUtil.java [hunk 39]
```diff
@@ -0,0 +1,45 @@
...
+    private static final String DELIMITER_REPLACEMENT = "_";
+
+    public boolean isMetricApplicable(final MetricConfig metricConfig, final String queueName) {
+        if (metricConfig == null) {
+            return false;
+        }
+        if (metricConfig.isEnabledForAll()) {
+            return true;
+        }
+        return metricConfig.getEnabledForQueues() != null
```
[the remark this answers is not in the document]
Comment (CyberDrudge) [reply src/main/java/io/appform/dropwizard/actors/metrics/MetricUtil.java:22]: No, do we need this to be dynamic?

File: src/main/java/io/appform/dropwizard/actors/common/RMQOperation.java [hunk 40]
```diff
@@ -0,0 +1,6 @@
+package io.appform.dropwizard.actors.common;
+
+public enum RMQOperation {
```
[the remark this answers is not in the document]
Comment (CyberDrudge) [reply src/main/java/io/appform/dropwizard/actors/common/RMQOperation.java:3]: Yes, we might add additional handling in the future for sideline queues etc.

File: src/main/java/io/appform/dropwizard/actors/RabbitmqActorBundle.java [hunk 41]
```diff
@@ -38,6 +44,9 @@ public abstract class RabbitmqActorBundle<T extends Configuration> implements Co
 
     @Getter
     private ConnectionRegistry connectionRegistry;
+    @Getter
+    private RMQObserver rootObserver;
```
[the remark this answers is not in the document]
Comment (CyberDrudge) [reply src/main/java/io/appform/dropwizard/actors/RabbitmqActorBundle.java]: Made it local variable

File: src/main/java/io/appform/dropwizard/actors/common/RMQOperation.java (hunk 40 above)
[the remark this answers is not in the document]
Comment (CyberDrudge) [reply src/main/java/io/appform/dropwizard/actors/common/RMQOperation.java]: Removed Enum class, used static string for Metric prefix

File: src/main/java/io/appform/dropwizard/actors/metrics/MetricUtil.java (hunk 39 above)
[the remark this answers is not in the document]
Comment (CyberDrudge) [reply src/main/java/io/appform/dropwizard/actors/metrics/MetricUtil.java:22]: Will use dynamic config

**Review by santanusinha — APPROVED** [review]

**Pull request merged**